### ISSUE TITLE
Add support for JavaScript-based transformation

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/kinesis-tee#9]
@ninjabear - I think we should add this into the 0.1.0:

Replace:

```
"transformer": "SNOWPLOW_TO_JSON", Or null
```

with:

```
"transformer": {
  "builtIn": "SNOWPLOW_TO_JSON"
} // union with
"transformer": null
  // union with
"transformer": {
  "javascript": "BASE64 ENCODED STRING"
}
```

where the javascript must be a function:

``` javascript
function transform(record) {
    return record; // identity
}
```

This will instantly make this Kinesis Tee much more powerful. What do you reckon?

**Comment by ninjabear** [comment]
Hey @alexanderdean ! I like it! The only question I have is, if you're using this with Snowplow you'd probably want to run both the `SNOWPLOW_TO_JSON` transformer and your custom js afterwards right? So we'd need a way to chain transformations together?

Perhaps we can schedule this in an 0.2.0 after we get some feedback? Is there a specific use case you have in mind?

**Comment by alexanderdean** [comment]
Hey @ninjabear - I think for non-Snowplow users just having it without chaining would still be useful (it would also be useful for Snowplow users who want to manipulate the enriched events _as a TSV_) - but I take your point, this needs a bit more cooking. Let's push back to 0.2.0... I still think we should add the `builtIn` bit though to make the transformer a valid Avro union for the future, will update the README now...

**Comment by alexanderdean** [comment]
Actually, @ninjabear was right - it's important to allow chaining of these transformations from the 0.2.0 onwards. Because it's valuable to e.g. use `SNOWPLOW_TO_FLATTENED_JSON` (#10) and then alias some column names.

**Comment by alexanderdean** [comment]
We need chaining, and this should be done first in 0.2.0.

**Comment by alexanderdean** [comment]
@yalisassoon will use this to test the issues seen in Kinesis Analytics, hence why this should be done first...

**Comment by alexanderdean** [comment]
This should be done first as a blocker on @yalisassoon's R&D

**Comment by nakulgan** [comment]
I ran into a blocker with the config proposed by @alexanderdean. Avro4s does not play well with multiple named records in a Union. 

```
unions not yet supported beyond nullable fields
```

So I propose the following change for the config:

``` json
[{
    "name": "transformer",
    "type": [
        "null", {
            "name": "Transformer",
            "type": "record",
            "fields": [{
                "name": "type",
                "type": "string"
            }, {
                "name": "value",
                "type": "string"
            }]
        }
    ]
}, {
    "name": "filter",
    "type": [
        "null", {
            "name": "Filter",
            "type": "record",
            "fields": [{
                "name": "type",
                "type": "string"
            }, {
                "name": "value",
                "type": "string"
            }]
        }
    ]
}]
```

Example:

``` json
{
  "schema": "iglu:com.snowplowanalytics.kinesistee.tbd",
  "data": {
    "name": "My Kinesis Tee example",
    "targetStream": {
      "name": "my-target-stream",
      "targetAccount": null
    },
    "transformer": {
      "com.snowplowanalytics.kinesistee.config.Transformer": {
        "type": "BuiltIn",
        "value": "SNOWPLOW_TO_NESTED_JSON"
      }
    },
    "filter": null
  }
}
```

The Transform/Filter records could be an array of their corresponding type, which would play well with chaining transformers & filters by composing them into a pipeline. (see #18)

Thought @snowplow/ed-lewis @alexanderdean

**Comment by alexanderdean** [comment]
Hey @nakulgan - first off, I think confusingly the ticket description and the README to this project are both out-of-date, sorry about that. If you look at @ninjabear 's wiki page, which should be up-to-date:

https://github.com/snowplow/kinesis-tee/wiki/Guide-for-devops-users#setting-up-the-dynamodb-configuration-table

``` json
    "transformer": {
      "com.snowplowanalytics.kinesistee.config.Transformer": {
        "builtIn": "SNOWPLOW_TO_NESTED_JSON"
      }
    },
```

I'm not clear on why we have to break the configuration down further into:

``` json
    "transformer": {
      "com.snowplowanalytics.kinesistee.config.Transformer": {
        "type": "BuiltIn",
        "value": "SNOWPLOW_TO_NESTED_JSON"
      }
    },
```

Could you clarify?

**Comment by nakulgan** [comment]
Yes @alexanderdean Turns out Avrohugger does not support unions with more than two records.
See [here](https://github.com/julianpeeters/avrohugger/blob/master/avrohugger-core/src/main/scala/format/specific/converters/ScalaConverter.scala#L97)

From the Avro docs, it does support multiple named records, as in:

``` json
 {
         "name":"transformer",
         "type":[
             "null",
             {
                "name":"Transformer",
                "type":"record",
                "fields":[
                    {
                        "name":"builtIn",
                        "type": { "name": "BuiltIn", "type": "enum", "symbols": ["SNOWPLOW_TO_NESTED_JSON"] }
                    }
                ]
             },
             {
                 "name":"JavascriptTransformer",
                 "type":"record",
                 "fields":[
                     {
                         "name":"javascript",
                         "type": "string"
                     }
                 ]
              }
         ]
      }
```

which is what we need. This throws the runtime exception

```
unions not yet supported beyond nullable fields
```

Does it make sense to open an issue with Avrohugger?

**Comment by alexanderdean** [comment]
Thanks for clarifying! I reckon we open an issue with Avrohugger and see how big a job they think it is? Perhaps we can do the heavy lifting if it's not a huge refactor...

**Comment by julianpeeters** [comment]
Hi @nakulgan and @alexanderdean,

Supporting unions in full is certainly on the radar, but the blocker is that IMHO there is currently no suitable representation of unions in Scala, and so I'm eagerly, but stubbornly, awaiting dotty's union types to be included in Scala.

Obviously dotty is still sometime off, and other representations have been proposed, but I can't see a path to adopting either (well, without first establishing a **friendly**, **standard** default):
1. Shapeless unions - would be nice to support, but, as a standard representation of unions in avrohugger, it violates the principle of least power, locks in users to another dependency, and scares away contributors
2. `Any` - Since Java Avro uses `Object` to represent unions, `Any` was proposed, however have personal qualms with being responsible for a tool that generates unsafe code. I would not like to inherit a codebase full of large schemas/case classes full of Any.
3. Either 1 or 2, but sitting behind the current `Option[T]` strategy - a complication/technical debt/trap in some use-cases.

While I do warn that I'll be stubborn about unions, I'm certainly open to suggestions, and I don't discourage a fork. E.g., [RelateIQ](https://github.com/relateiq/avrohugger) seems to have a higher tolerance for `Any`.

**Comment by alexanderdean** [comment]
Hi @julianpeeters - thanks for the detailed explanation. It feels like first class language support for union types is going to be the killer feature for Scala.next, given how mainstream schema tech like Avro and JSON Schema has made unions...

Given the above, I reckon @nakulgan's suggestion of a `map[string, string]` is the way to go... @ninjabear thoughts?

**Comment by ninjabear** [comment]
You can limit the use of `Any` a bit by making a kind of marker trait e.g.

``` scala
sealed trait MyUnion {}
case class TypeA extends MyUnion
case class TypeB extends MyUnion
```

Collections from that point have the type `MyUnion` but can be partitioned/matched against their sub-types in a type-safe way.

I think maps is a legitimate way to go though, especially if making a change like above would be very time consuming or has other problems.

**Comment by julianpeeters** [comment]
Thanks for the good suggestions, I'll be sure to include them in any future discussions.

Avrohugger supports both data-first (from schemas) and code-first (expanding existing Scala code to implement `SpecificRecordBase`) use cases, and both `Map` and ADTs are already used to represent Avro features.
- `Map` represents `MAP` of course, and since `MAP` and `UNION` are both datatypes, Avrohugger would be unable to decide which to Avro type it represented in code-first situations
- ADTs are used to represent protocols, and so it does indeed seem possible to search the compilation set to see if it's used as a field type in order to disambiguate protocols from unions

So, while I don't want to raise any hopes, the `sealed trait` approach looks promising. I need to think a bit more about any pitfalls this may introduce, particularly in terms of schema/protocol evolutions, and if that looks ok I'll solicit some feedback from Avrohugger's gitter channel.

**Comment by alexanderdean** [comment]
Thanks @julianpeeters !
